apps/performance-tests/src/StatelessHierarchyProvider.ts

```diff
@@ -0,0 +1,135 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+/* eslint-disable @itwin/no-internal */
+
+import { expand, filter, from, mergeAll, of, tap } from "rxjs";
+import { IModelDb } from "@itwin/core-backend";
+import { DbQueryRequest, DbQueryResponse, DbRequestExecutor, DbRequestKind, ECSqlReader } from "@itwin/core-common";
+import { ISchemaLocater, Schema, SchemaContext, SchemaInfo, SchemaKey, SchemaMatchType } from "@itwin/ecschema-metadata";
+import { createECSqlQueryExecutor, createMetadataProvider } from "@itwin/presentation-core-interop";
+import { createLimitingECSqlQueryExecutor, HierarchyNode, HierarchyProvider } from "@itwin/presentation-hierarchy-builder";
+import { ModelsTreeDefinition } from "@itwin/presentation-models-tree";
+
+const ENABLE_SCHEDULED_NODES_LOGGING = false;
+const ECSQL_ROW_LIMIT = 1000;
+
+export class StatelessHierarchyProvider {
+  private readonly _provider: HierarchyProvider;
+
+  constructor(
+    iModelDb: IModelDb,
+    private readonly _nodeRequestLimit = 10,
+  ) {
+    this._provider = createProvider(iModelDb);
+  }
+
+  public async loadInitialHierarchy(): Promise<void> {
+    await this.loadNodes(StatelessHierarchyProvider.initialHasChildren);
+  }
+
+  public async loadFullHierarchy(): Promise<void> {
+    await this.loadNodes(StatelessHierarchyProvider.fullHasChildren);
+  }
+
+  private static initialHasChildren(node: HierarchyNode): boolean {
+    return StatelessHierarchyProvider.fullHasChildren(node) && !!node.autoExpand;
+  }
+
+  private static fullHasChildren(node: HierarchyNode): boolean {
+    return node.children === true || (Array.isArray(node.children) && node.children.length > 0);
+  }
```

Comment:
These static functions being passed to `loadNodes` are very unclear to me. Why not just inline them? E.g.
```ts
  public async loadInitialHierarchy(): Promise<void> {
    await this.loadNodes((parent) => !!parent.children);
  }
```
Note: `Hierarchy.children` at was defined as `boolean | HierarchyNode[]` at the point the tests were written. Now it's always a `boolean`, so no need to check for array type here.

Reply:
Yes, I copied this long predicate from the load tests which seemed weird to me but I thought that there's probably a reason for that 😄